feat: add app.getSystemLocale() method (#35697)
electron/electron: docs/api/app.md

**Note:** When distributing your packaged app, you have to also ship the
`locales` folder.

**Note:** On Windows, you have to call it after the `ready` event is emitted.

### `app.getLocaleCountryCode()`

Returns `string` - User operating system's locale two-letter [ISO 3166](https://www.iso.org/iso-3166-country-codes.html) country code. The value is taken from native OS APIs.

**Note:** When unable to detect locale country code, it returns empty string.

### `app.getSystemLocale()`

Returns `string` - The current system locale, fetched using Chromium's `l10n_util` library.
Possible return values are documented [here](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc).

> **Note:** This API must be called after the `ready` event is emitted.

Reviewer: Can you update the `app.getLocale` note to match this one?

Author: The `app.getLocale` API is older and doesn't have a check to ensure the caller is calling it after the `ready` event. The downside of adding the check is it might break platforms where one can get the correct result even before the `ready` event, but the upside is it could prevent [a few bug reports](https://github.com/electron/electron/issues?q=is%3Aissue+getLocale+is%3Aclosed+ready). If it is worth adding the check, I think that should go into a separate PR.

Reviewer: I would still suggest updating the docs to state that the API must be called after the `ready` event, even if that isn't strictly enforced on macOS and Linux currently. I agree, though, that adding an enforcement check would be better to do in another PR.

**Note:** When unable to detect locale country code, it returns empty string.

### `app.getSystemLocale()`

Returns `string` - The current system locale, fetched using Chromium's `l10n_util` library.

> Possible return values are documented [here](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc).

Reviewer: Also, looking at the docs for `app.getLocale`, it's not clear to me what it returns if you don't specify the `--lang` flag explicitly. Does it fall back to the system locale or just default to English? That probably would have been good to document before, but while we're adding `app.getSystemLocale` it would be good to clarify that default behavior of `app.getLocale`. If it returns the same value as `app.getSystemLocale`, it would probably be sufficient to note that fact and link to the `app.getSystemLocale` section.

> Possible return values are documented [here](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc).

Author: For reference, `app.getLocale` calls into [l10n_util::GetApplicationLocale](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc;drc=4c33ea7ce9a4064087f939723c0e9713193cf54f;l=573) which is much more involved.
All I know for now is that it might not return the same value as `app.getSystemLocale` by default. On my Windows machine, it returns the same value as `app.getSystemLocale`, but it seems like it can fall back to `en-US` on [GetApplicationLocaleInternalMac](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc;l=498;drc=4c33ea7ce9a4064087f939723c0e9713193cf54f).

> Possible return values are documented [here](https://source.chromium.org/chromium/chromium/src/+/main:ui/base/l10n/l10n_util.cc).

Reviewer 2: :+1: on @itsananderson's suggestion of finding a version-friendly way of documenting the values for this new function and for `app.getLocale()` but it's out-of-scope for this PR

> **Note:** On Windows, you have to call it after the `ready` events gets emitted.

Reviewer 3: Lets be consistent for all platforms and only allow calling this API after app ready. Also the API should throw an error if called before app ready to support this note. An example, https://github.com/electron/electron/blob/fa3cd1747556b66a3e714f99e988e47334187e44/shell/browser/api/electron_api_app.cc#L1189-L1193

> **Note:** On Windows, you have to call it after the `ready` events gets emitted.

Reviewer 3: ```suggestion
**Note:** This API must be called after the `ready` event is emitted.
```
